cosmwasm/ibc-union/lightclient/attested/src/contract/execute.rs

```diff
@@ -0,0 +1,227 @@
+use std::{collections::BTreeMap, num::NonZero};
+
+use cosmwasm_std::{DepsMut, Event, Response, StdError, ensure};
+use depolama::StorageExt;
+use unionlabs::{
+    encoding::{Bincode, EncodeAs},
+    primitives::{H256, H512},
+};
+
+use crate::{
+    errors::Error,
+    state::{
+        AttestationAttestors, Attestations, Attestors, HeightTimestamps, PendingAttestations,
+        Quorum,
+    },
+    types::{Attestation, AttestationKey},
+};
+
+pub fn attest(
+    mut deps: DepsMut,
+    attestation: Attestation,
+    attestor: H256,
+    signature: H512,
+) -> Result<Response, Error> {
+    deps.storage
+        .maybe_read::<Attestors>(&(attestation.chain_id.clone(), attestor))?
+        .ok_or_else(|| Error::InvalidAttestor {
+            chain_id: attestation.chain_id.clone(),
+            attestor,
+        })?;
+
+    if let Some(previously_attested_timestamp) = deps
+        .storage
+        .maybe_read::<HeightTimestamps>(&(attestation.chain_id.clone(), attestation.height))?
+        && previously_attested_timestamp != attestation.timestamp
+    {
+        return Err(Error::InconsistentTimestamp {
+            chain_id: attestation.chain_id.clone(),
+            height: attestation.height,
+            timestamp: attestation.timestamp,
+            previously_attested_timestamp,
+        });
+    }
+
+    let attestation_key = AttestationKey {
+        chain_id: attestation.chain_id.clone(),
+        height: attestation.height,
+        key: attestation.key.clone(),
+    };
+
+    if let Some(value) = deps.storage.maybe_read::<Attestations>(&attestation_key)? {
+        return Err(Error::AlreadyAttested {
+            chain_id: attestation.chain_id.clone(),
+            height: attestation.height,
+            timestamp: attestation.timestamp,
+            key: attestation.key,
+            value,
+        });
+    }
+
+    ensure!(
```

Comment:
considering we might have gud amount of traffic, maybe we can change the api to be able to verify multiple attestations at once through `cosmwasm`'s batch verification api.

Reply:
i considered this, it would be a trivial addition down the line on the contract side but it will be a decent amount more work on the voyager side so i decided to just go with this MVP for now